source/subscription-manager/subscription_manager.c

```diff
@@ -0,0 +1,168 @@
+
+/**
+ * @brief Maximum number of subscriptions maintained by the subscription manager
+ * simultaneously.
+ */
+#ifndef SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS
+    #define SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS    ( 10U )
+#endif
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief An element in the list of subscriptions maintained by the agent.
+ *
+ * @note The agent allows multiple tasks to subscribe to the same topic.
+ * In this case, another element is added to the subscription list, differing
+ * in the intended publish callback.
+ */
+typedef struct subscriptionElement
+{
+    IncomingPubCallback_t pxIncomingPublishCallback;
+    void * pvIncomingPublishCallbackContext;
+    uint16_t uFilterStringLength;
+    const char * pcSubscriptionFilterString;
+} SubscriptionElement_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The static array of subscription elements.
+ *
+ * @note No thread safety is required to this array, since the updates the array
+ * elements are done only from one task at a time.
+ */
+static SubscriptionElement_t xSubscriptionList[ SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS ];
+
+/*-----------------------------------------------------------*/
+
+BaseType_t addSubscription( const char * pcTopicFilterString,
+                            uint16_t uTopicFilterLength,
+                            IncomingPubCallback_t pxIncomingPublishCallback,
+                            void * pvIncomingPublishCallbackContext )
+{
+    int32_t lIndex = 0;
+    size_t xAvailableIndex = SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS;
+    BaseType_t xReturnStatus = pdFALSE;
+
+    /* Start at end of array, so that we will insert at the first available index.
+     * Scans backwards to find duplicates. */
+    for( lIndex = ( int32_t ) SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS - 1; lIndex >= 0; lIndex-- )
+    {
+        if( xSubscriptionList[ lIndex ].uFilterStringLength == 0 )
+        {
+            xAvailableIndex = lIndex;
+        }
+        else if( ( xSubscriptionList[ lIndex ].uFilterStringLength == uTopicFilterLength ) &&
+                 ( strncmp( pcTopicFilterString, xSubscriptionList[ lIndex ].pcSubscriptionFilterString, uTopicFilterLength ) == 0 ) )
+        {
+            /* If a subscription already exists, don't do anything. */
+            if( ( xSubscriptionList[ lIndex ].pxIncomingPublishCallback == pxIncomingPublishCallback ) &&
+                ( xSubscriptionList[ lIndex ].pvIncomingPublishCallbackContext == pvIncomingPublishCallbackContext ) )
+            {
+                LogWarn( ( "Subscription already exists.\n" ) );
+                xAvailableIndex = SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS;
+                xReturnStatus = pdTRUE;
+                break;
+            }
+        }
+    }
+
+    if( ( xAvailableIndex < SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS ) && ( pxIncomingPublishCallback != NULL ) )
```

Comment:
Should the `pxIncomingPublishCallback` be asserted against `NULL` as it seems like a required parameter for registering the callback? 
Also, it would be better to avoid searching for an available index is the callback is `NULL`

Reply:
The NULL check shouldn't be necessary any more since the auto reconnect resubscription code was removed